### PR TITLE
Remove block tools and keychain globals

### PR DESCRIPTION
**Pull request by aqk**
This patch removes a number of global variables from the tests. This builds on an earlier patch from @paninaro

### CODE REVIEW

File: tests/core/daemon/test_daemon.py
Comment (aqk) [inline tests/core/daemon/test_daemon.py:56]: Please take a close look at this line:

`new_config["daemon_port"] = 55401`

File: tests/core/daemon/test_daemon.py
Comment (aqk) [inline tests/core/daemon/test_daemon.py:53]: Please take a close look at `get_b_tools`

**Comment by aqk** [comment]
@paninaro could you take a look at `tests/core/cmds/test_keys.py::TestKeysCommands::test_generate_with_new_config` ?

````FAILED                                                                           [1/1]

_________ TestKeysCommands.test_generate_with_new_config ____
tests/core/cmds/test_keys.py:133: in test_generate_with_new_config
    assert len(keychain.get_all_private_keys()) == 1
E   assert 0 == 1
```

File: tests/core/test_daemon_rpc.py
Comment (aqk) [inline tests/core/test_daemon_rpc.py:2]: Python fixture auto-imports are not recognized by flake / pylint / black

File: setup.py
Comment (aqk) [inline setup.py:46]: Older `pytest-asyncio` is missing features we use.

File: tests/block_tools.py
Comment (aqk) [inline tests/block_tools.py:2023]: We can remove these two print stmts before this patch lands

File: tests/blockchain/test_blockchain_transactions.py
Comment (aqk) [inline tests/blockchain/test_blockchain_transactions.py:17]: There are a number of warts & limitations in the Python fixtures implementation -
Removing these explicit references and putting them in `conftest.py` is the accepted way of using fixtures.

**Comment by aqk** [comment]
Important files:

```
tests/setup_nodes.py
tests/block_tools.py
tests/conftest.py
chia/util/keyring_wrapper.py
```

File: tests/core/daemon/test_daemon.py
Comment (arvidn) [reply tests/core/daemon/test_daemon.py:56]: In my patch (which I'm hoping to land before this) I remove this assignment and make every `create_blockchain_tools` pick an available port and use that.

File: chia/util/keyring_wrapper.py
Comment (arvidn) [inline chia/util/keyring_wrapper.py:112]: I thought the main reason to have a `create()` function (instead of a constructor) is that it can be `async`. But this isn't. @altendky has also pointed out that with a `create()` function you can mark the class as `@dataclass` (which has some benefits I haven't quite understood yet). But this class isn't a `dataclass` either.

File: chia/util/keyring_wrapper.py
Comment (arvidn) [inline chia/util/keyring_wrapper.py:117]: is `cls` sometimes not `KeyringWrapper`?
If it always is, I think it would be clearer to just construct that type explicitly.

File: chia/util/keyring_wrapper.py
Comment (arvidn) [inline chia/util/keyring_wrapper.py:210]: I don't see any calls to this function. It seems like a risky hole in the abstraction too. Can it be removed?

File: tests/block_tools.py
Comment (arvidn) [reply tests/block_tools.py:2023]: I take it you intend to remove those global counters too, right?

File: tests/block_tools.py
Comment (arvidn) [inline tests/block_tools.py:2014]: yes, they do. They all need different root directories containing different `config.yaml` files. The daemon's listen port is configured in there, and the only way a test can control which listen port it uses is to write it to the config file.

File: tests/blockchain/test_blockchain_transactions.py
Comment (arvidn) [reply tests/blockchain/test_blockchain_transactions.py:17]: currently `bt` is a global variable that's being imported here. in the new world, `bt` is a fixture, so I don't think you *can* import it (you certainly don't want to)

File: tests/conftest.py
Comment (arvidn) [inline tests/conftest.py:22]: I think you will need to make this a generator fixture, otherwise you don't get an opportunity to clean it up (which is critical).

Also, one of the problems with the current global variable `bt` is that it's mutable and tests *do* mutate it. Giving this "session" scope, preserves some of those problems (it's still mutable, right?). Would it be feasible to at least make it "module" scoped?

File: tests/conftest.py
Comment (arvidn) [inline tests/conftest.py:18]: this should also be a fixture, but I consider making `bt` a fixture the top priority. The global keyring can be cleaned up later

File: tests/conftest.py
Comment (arvidn) [inline tests/conftest.py:34]: I don't see this used anywhere. It also looks suspicious, now that `bt` is a fixture, it doesn't look right to also create one like this (that also may not be cleaned up correctly on shutdown).

Is it possible to *set* the keychain on an existing `bt` object? If so, tests that need a specific keyring could still use the `bt` fixture.

File: tests/connection_utils.py
Comment (arvidn) [inline tests/connection_utils.py:27]: does the hostname we claim to have matter for any test?
It might not, in which case we might be able to just pick a random string here and be done with it.

File: tests/core/daemon/test_daemon.py
Comment (arvidn) [inline tests/core/daemon/test_daemon.py:32]: I'm pretty sure `get_b_tools` is currently referring to the global `bt` object (and you could replace that with the new fixture. However, `get_b_tools_1` should be a copy of `bt`. i.e. the same kind of fixture but a separate instance.

File: tests/connection_utils.py
Comment (arvidn) [reply tests/connection_utils.py:27]: I'm experimenting with simplifying this here: https://github.com/Chia-Network/chia-blockchain/pull/10371

File: tests/core/test_daemon_rpc.py
Comment (arvidn) [reply tests/core/test_daemon_rpc.py:2]: you're removing these, right?

File: tests/util/blockchain.py
Comment (arvidn) [inline tests/util/blockchain.py:38]: these comments should be removed before we land (and maybe experimented with further in a separate PR)

File: tests/util/keyring.py
Comment (arvidn) [inline tests/util/keyring.py:169]: and this comment

**Review by arvidn — CHANGES_REQUESTED** [review]
I can't wait for this to land! However, you may want to rebase on top of https://github.com/Chia-Network/chia-blockchain/pull/10307

There's an opportunity for simplifying this patch by also applying https://github.com/Chia-Network/chia-blockchain/pull/10371 first too

File: tests/conftest.py
Comment (aqk) [reply tests/conftest.py:35]: That is a really good idea. I will add a comment.
However, I think the keychain requires that there is only one instance at the moment.

File: tests/conftest.py
Comment (aqk) [reply tests/conftest.py]: Thanks, I knew I'd need that once I added the keychain fixture. Also, I looked, but I'm not aware of any BlockTools cleanup functions. Let me look at that again.

File: chia/util/keyring_wrapper.py
Comment (aqk) [reply chia/util/keyring_wrapper.py:112]: You know, I had assumed that the reason for the create method was so that we could avoid a global, and instead construct at a later time of our choosing, but I don't know. @altendky @paninaro ?

File: chia/util/keyring_wrapper.py
Comment (aqk) [reply chia/util/keyring_wrapper.py:117]: Yeah, I am on the fence - naming the expected type seems better, because it is more constrained.
I'm not sure if the `cls()` helper does some extra Python class magic that I'm not aware of.

File: chia/util/keyring_wrapper.py
Comment (aqk) [reply chia/util/keyring_wrapper.py:210]: Oh, if this is added by my patch, this is probably leftover code - I'll remove it.

File: tests/core/daemon/test_daemon.py
Comment (aqk) [reply tests/core/daemon/test_daemon.py:28]: Yes, that was my impression - `get_b_tools_1` is safe as-is for now, correct?

File: chia/util/keyring_wrapper.py
Comment (paninaro) [reply chia/util/keyring_wrapper.py:112]: I honestly don't remember why I had added the create method here... Looking at where it's used and what's checked into main, I don't see any functional difference between `create()` and `__init__()`. It may have been that I was trying to convey that multiple KeyringWrappers shouldn't be created.

File: tests/block_tools.py
Comment (aqk) [reply tests/block_tools.py:2041]: I'll remove these when they are not being run multiple times

File: chia/util/keyring_wrapper.py
Comment (aqk) [reply chia/util/keyring_wrapper.py]: removed

File: tests/connection_utils.py
Comment (aqk) [reply tests/connection_utils.py:27]: Ya, I initially had it that way with the global, too. Thanks!

File: tests/block_tools.py
Comment (mariano54) [inline tests/block_tools.py:2055]: why do we need these globals?

**Review by mariano54 — COMMENTED** [review]
Seems reasonable to me. I have not reviewed the keychain changes.

File: tests/block_tools.py
Comment (aqk) [reply tests/block_tools.py:2055]: We don't! I just want to see them go to 1 before we remove these two functions entirely.

File: chia/util/keyring_wrapper.py
Comment (aqk) [reply chia/util/keyring_wrapper.py:117]: @altendky do you have an opinion? In this case, I would tend to go with the more restrictive type, a Arvid is suggesting, but the usual thing in Python seems to be `cls()` - perhaps to avoid a class rename breaking code?

File: tests/block_tools.py
Comment (aqk) [reply tests/block_tools.py:2055]: They allow us to print how many times the BlockTools create functions have been called.
They aren't going to stay.

**Review by paninaro — DISMISSED** [review]
Looks good to me

File: tests/core/util/test_streamable.py
Comment (altendky) [inline tests/core/util/test_streamable.py:371]: Did this need to move?

File: tests/wallet/rpc/test_wallet_rpc.py
Comment (altendky) [inline tests/wallet/rpc/test_wallet_rpc.py:142]: Do we want to be switching to pulling this out of the config?  If so, should we be using the fixture up above?

**Review by altendky — COMMENTED** [review]
Thanks for all the work!  Looks like we're close...

File: chia/util/keyring_wrapper.py
Comment (paninaro) [reply chia/util/keyring_wrapper.py:218]: I would suggest that we resist the urge to make additional structural changes at this point. Let's let this PR land, get coverage to 100% and then revisit :). If it ain't broke and all...

File: tests/util/keyring.py
Comment (arvidn) [inline tests/util/keyring.py:168]:
```suggestion
        keychain.keyring_wrapper = KeyringWrapper.create(keys_root_path=Path(temp_dir))
        assert keyring_wrapper is not None
        keychain.keyring_wrapper = keyring_wrapper
```

Comment (arvidn) [reply tests/util/keyring.py:168]: this is to fix the mypy error (untested):
```
tests/util/keyring.py:168: error: Incompatible types in assignment (expression has type "Optional[KeyringWrapper]", variable has type "KeyringWrapper")  [assignment]
```

My understanding is that this is how you unwrap optionals in python

File: chia/util/keyring_wrapper.py
Comment (arvidn) [reply chia/util/keyring_wrapper.py:111]: It would seem reasonable to not return an `Optional[]`, but just a `_T_KeyringWrapper` directly. If there's an error, throw an exception. This would fix the mypy error

File: tests/util/keyring.py
Comment (altendky) [reply tests/util/keyring.py:168]: `.create()` is only optional by mistake I think.  https://github.com/Chia-Network/chia-blockchain/pull/10358#discussion_r820005241  If you are sprinkling `assert`s around for `None`s, maybe you didn't want the thing to be optional.  Also see the keybase discussion about `assert`s.  `:]`

Comment (arvidn) [reply tests/util/keyring.py:168]: I agree that `create()` shouldn't return an `Optional`, but assuming the existing code in this patch is correct, this would be the correct use of assert. The existing code clearly expects it to never return `None` in this case. If it did, it would be a programmer mistake, not an expected/recoverable failure.

Comment (aqk) [reply tests/util/keyring.py]: Thanks guys.

File: tests/core/util/test_streamable.py
Comment (aqk) [reply tests/core/util/test_streamable.py:371]: Yes! When executed as a part of the class, it broke because of interactions stemming from `TestStreamable` inheriting from `unittest.TestCase`

File: tests/wallet/rpc/test_wallet_rpc.py
Comment (aqk) [reply tests/wallet/rpc/test_wallet_rpc.py:142]: `self_hostname` and `bt.config["hostname"]` are certain to be the same if a test imports both `bt` and `self_hostname`

**Comment by arvidn** [comment]
@aqk the submodule update is a mistake I think

File: tests/core/util/test_streamable.py
Comment (altendky) [reply tests/core/util/test_streamable.py:371]: Oh, right...  it grew a fixture dependency.  Maybe we should convert this from unittest to pytest fully as a pre-PR.

**Review by arvidn — CHANGES_REQUESTED** [review]
I think there are a still a lot of remnants left from the first version of this patch that can (and should) be removed.

File: tests/wallet/rpc/test_wallet_rpc.py
Comment (altendky) [reply tests/wallet/rpc/test_wallet_rpc.py:142]: Which is the preferred method?  It would seem like the fixture would be since that would maintain consistency.  Also, it would leave these lines as they were before.  The other lines should be changed to match, but since this could all not be changed at all in this PR, I'm fine with that route.

File: tests/core/util/test_streamable.py
Comment (altendky) [reply tests/core/util/test_streamable.py:371]: I'm happy to write it.

File: chia/util/keyring_wrapper.py
Comment (altendky) [reply chia/util/keyring_wrapper.py]: Looks like all changes in this file were removed.

File: tests/core/test_daemon_rpc.py
Comment (aqk) [reply tests/core/test_daemon_rpc.py:14]: We may not need it in the sense of "is this necessary". But fixing this would be a further improvement, and out of the (massive) scope of this patch, because `setup_daemon(btools)` in `tests/setup_nodes.py` already too a BlockTools as an argument. So, we either have to understand and change `setup_daemon`, or we have toadd a parameter to `get_daemon` to compensate for the global that has been removed.

File: tests/core/ssl/test_ssl.py
Comment (aqk) [reply tests/core/ssl/test_ssl.py:24]: I don't see how we can get away without passing in `self_hostname` to `establish_connection` - it is used twice in the body. `self_hostname` just used to be a global that silently brought state into every test method in a module that imported it.

File: tests/setup_nodes.py
Comment (aqk) [reply tests/setup_nodes.py:95]: I agree that I am setting daemon_port too many times in this patch. Definitely with you there.

However, it gives me way more pain that state is being passed downstream through the config file, so you can get into situations like:

```
def fun_a():
  bt.config["obscure_attribute"] = whatever
  fun_b(bt)

def fun_b(bt):  # e.g. setup_xyz
  fun_c(bt)

def fun_c(bt):
  use_obscure_attribute(bt)
```

So, making a change in fun_a can break the call at fun_c

I suppose you could argue that now we are at least tracking the passage of BlockTools through formal parameters, so I could be convinced to back out the redundant but it still bugs me that state that we NEED is not being pass in the formal parameters.

This is one of the things that makes Blocktools so opaque, hard to understand, and debug.

File: tests/setup_nodes.py
Comment (aqk) [reply tests/setup_nodes.py:116]: Right. This is a broken way of doing things... I didn't introduce doing it this way, but maybe it can be taken out now that your patch is in.

Comment (aqk) [reply tests/setup_nodes.py:116]: I'll test and see if I can take this out!

File: tests/block_tools.py
Comment (aqk) [reply tests/block_tools.py:136]: Yes, now that this is debugged, that can go.

File: tests/core/daemon/test_daemon.py
Comment (aqk) [reply tests/core/daemon/test_daemon.py:35]: Yes. This was before your `find_available_listen_port` patch.

Comment (aqk) [reply tests/core/daemon/test_daemon.py:35]: Code leftovers removed

File: setup.py
Comment (aqk) [reply setup.py:47]: #10560 approved

File: tests/core/ssl/test_ssl.py
Comment (arvidn) [reply tests/core/ssl/test_ssl.py:24]: they way we do in `main` right now. Import it.

Comment (arvidn) [reply tests/core/ssl/test_ssl.py:24]: I'm suggesting you revert this part of your patch. to import `self_hostname` in all places it's used, rather than importing it in *some* places and then pass it in as a new parameter

File: tests/setup_nodes.py
Comment (arvidn) [reply tests/setup_nodes.py:116]: github shows this code as new in your patch to me.

File: tests/setup_nodes.py
Comment (arvidn) [reply tests/setup_nodes.py:95]: in my mind, the daemon port (and most configs actually) should be treated as immutable. If we do that (and are suspicious of any place that alters the config, I think we're good.

With that view, configuring the daemon port in the configs once when the `BlockTools` object is created seems the most reasonable.

File: tests/conftest.py
Comment (arvidn) [inline tests/conftest.py:31]: oh, I see. You made this a fixture. I think that's a mistake. There's no upside and it adds a lot of extra code.

Comment (arvidn) [reply tests/conftest.py:31]: it also means that *any* test that starts a listen port and connects to it is *forced* to create the `BlockTools` fixture. That doesn't make any sense at all.

If you really think it should be a fixture, at least don't make it depend on `bt`. It's just a constant string literal.

File: tests/wallet/rpc/test_wallet_rpc.py
Comment (arvidn) [reply tests/wallet/rpc/test_wallet_rpc.py:142]: I think it's a mistake to make `self_hostname` a fixture. it restricts access to it to only test functions and other fixtures. Currently we use it in other places too.

It should especially not have to depend on constructing a whole `BlockTools` object.

Comment (arvidn) [reply tests/wallet/rpc/test_wallet_rpc.py:142]: > Which is the preferred method? It would seem like the fixture would be since that would maintain consistency.

in `main` (that this PR is targeting), `self_hostname` is a global string literal in the tests. I don't think there's a reason for this patch to change that.

**Comment by aqk** [comment]
Land after https://github.com/Chia-Network/chia-blockchain/pull/10577

File: tests/conftest.py
Comment (aqk) [reply tests/conftest.py:31]: Well, it's trues that none of the current tests rely on the value of `self_hostname`
It's also true that leaving it the way it is now invades countless test functions with a global.

File: tests/setup_nodes.py
Comment (aqk) [reply tests/setup_nodes.py]: Removed.

File: tests/core/test_daemon_rpc.py
Comment (aqk) [reply tests/core/test_daemon_rpc.py:14]: Take a look at the change history of `setup_daemon` to see what I mean.

Oh, wait. When you say "we don't need to do this anymore", do you mean:
`    bt._config["daemon_port"] = find_available_listen_port()` ?

I thought you meant adding `bt` as an argument.
Sure, I'll take a look at that config assignment - I don't think we need it.

Comment (aqk) [reply tests/core/test_daemon_rpc.py]: Okay, take a look at that now.

**Comment by aqk** [comment]
@arvidn Thanks, submodule updated

**Comment by aqk** [comment]
This is now rebased on top of https://github.com/Chia-Network/chia-blockchain/pull/10577

File: tests/conftest.py
Comment (arvidn) [reply tests/conftest.py:31]: a global *constant*, not variable. it's like naming pi, it won't change, it's a constant.

File: tests/setup_nodes.py
Comment (arvidn) [inline tests/setup_nodes.py:52]: here's the comment describing what `self_hostname` is.

**Review by arvidn — APPROVED** [review]

**Review by paninaro — APPROVED** [review]

**Review by wjblanke — APPROVED** [review]
aok

**Pull request merged**
